### PR TITLE
puredata: upgrade to 0.54-0 and enable darwin

### PR DESCRIPTION
**Pull request by carlthome**
###### Description of changes

Upgrade Pure Data to latest and fix Darwin support.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by carlthome** [comment]
@onny thanks! Think this is ready to go now. Seems to work fine in sandboxed Darwin at least, in manual testing.

I could manually test on x86 Linux (NixOS desktop) if needed too.

**Comment by carlthome** [comment]
Any action needed on my part to get this in?

**Comment by carlthome** [comment]
For anyone on macOS wanting to run Pure Data, feel free to use 

```
nix run github:carlthome/nixpkgs#puredata
```

in the meantime.

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 230569` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>helmholtz</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>cyclone</li>
    <li>gem</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>maxlib</li>
    <li>mrpeach</li>
    <li>puredata</li>
    <li>puremapping</li>
    <li>timbreid</li>
    <li>zexy</li>
  </ul>
</details>

**Comment by wegank** [comment]
Breaks `cyclone` and `gem`, both need to be updated in the same PR.

**Comment by carlthome** [comment]
cyclone cannot compile because it's using private functions from puredata: https://github.com/pure-data/pure-data/issues/2043

**Comment by carlthome** [comment]
And gem is having

```
Event.cpp: In function 'void dequeueEvents()':
Event.cpp:353:5: error: 'error' was not declared in this scope; did you mean 'perror'?
  353 |     error("dequeue NULL queue");
      |     ^~~~~
      |     perror
Event.cpp:358:5: error: 'error' was not declared in this scope; did you mean 'perror'?
  358 |     error("dequeue empty queue");
      |     ^~~~~
```

**Comment by carlthome** [comment]
Looks like bumping Gem to latest unstable made that build at least. Still need to make cyclone work though.

Would also be nice to see a stable release for Gem instead of calling the package name "gem-unstable": https://github.com/umlaeute/Gem/issues/371

**Comment by carlthome** [comment]
@mitchmindtree wanna collaborate on this? Figure next step is to resolve https://github.com/pure-data/pure-data/issues/2043#issuecomment-1675904751 in upstream.

**Comment by magnetophon** [comment]
FWIW: I also get:

Result of `nixpkgs-review pr 230569` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>cyclone</li>
  </ul>
</details>
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>gem</li>
    <li>helmholtz</li>
    <li>magnetophonDSP.VoiceOfFaust</li>
    <li>maxlib</li>
    <li>mrpeach</li>
    <li>puredata</li>
    <li>puremapping</li>
    <li>timbreid</li>
    <li>zexy</li>
  </ul>
</details>

**Comment by carlthome** [comment]
Until https://github.com/porres/pd-cyclone/issues/627 is resolved and a patch release exists, I think it's best to mark cyclone as broken since it doesn't compile. Regardless, it would be nice to get out an updated Pure Data version to have it work on Macs. Thoughts?

**Comment by carlthome** [comment]
Great news @magnetophon!

**Comment by carlthome** [comment]
So should we merge this now?

**Review by magnetophon — COMMENTED** [review]
Result of `nixpkgs-review pr 230569` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>cyclone</li>
    <li>gem</li>
    <li>helmholtz</li>
    <li>magnetophonDSP.VoiceOfFaust</li>
    <li>maxlib</li>
    <li>mrpeach</li>
    <li>puredata</li>
    <li>puremapping</li>
    <li>timbreid</li>
    <li>zexy</li>
  </ul>
</details>

Seemed fine, until I tried to test my most important pd patch, the one inside Voice Of Faust.

I got:
```
verbose(0): osc/packOSC
verbose(1): ... couldn't create
verbose(0): net/udpsend
verbose(1): ... couldn't create
fiddle version 1.1 TEST4
error: /nix/store/6gxvrxaql2kmf4vqmlnjwp0abr8w1nvr-VoiceOfFaust-1.1.4/PureData/helmholtz~.pd_linux:libstdc++.so.6: cannot open shared object file: No such file or directory
verbose(0): helmholtz~
verbose(1): ... couldn't create
verbose(0): net/udpreceive 5511
verbose(1): ... couldn't create
verbose(0): osc/unpackOSC
verbose(1): ... couldn't create
verbose(0): osc/pipelist
verbose(1): ... couldn't create
verbose(0): osc/routeOSC /VoiceOfFaust/OSC/max_pitch /VoiceOfFaust/OSC/min_pitch
verbose(1): ... couldn't create
Terminated
```

Not sure what causes that, I haven't had time to investigate.

**Comment by carlthome** [comment]
> Not sure what causes that, I haven't had time to investigate.

Hmm, on a skim it looks like there's a newer version of C++ assumed, since `libstdc++.so.6: cannot open shared object file: No such file or directory`

**Comment by raboof** [comment]
I tried the following shell with this branch (`nix-shell -I nixpkgs=/path/to/this/pr`):

```
{ pkgs ? import <nixpkgs> {} }:

pkgs.mkShell {
  buildInputs = [
    (pkgs.puredata-with-plugins [ pkgs.gem ])
  ];
}
```

starting `pd -lib Gem` gives me `Gem: can't load library`. On nixos-unstable loading Gem does start, but ultimately fails (citing not being able to load `libndi.so.4`). I'm not sure that's the same problem, but it makes me reluctant to merge this PR

**Comment by carlthome** [comment]
Thanks @raboof! I'll look into test running and fixing this when I get to a Linux workstation.

```
nix develop --impure --expr '{pkgs ? import <nixpkgs> {}, ...}: pkgs.mkShell { buildInputs = [ (pkgs.puredata-with-plugins [ pkgs.gem ]) ]; }'
```

Hoping it's just a matter of adding libndi to buildInputs.

File: pkgs/applications/audio/pd-plugins/gem/default.nix [hunk 1]
```diff
@@ -39,11 +41,11 @@ stdenv.mkDerivation rec {
...
 
-  meta = {
+  meta = with lib; {
     description = "Graphics Environment for Multimedia";
     homepage = "http://puredata.info/downloads/gem";
-    license = lib.licenses.gpl2Plus;
-    maintainers = [ lib.maintainers.raboof ];
-    platforms = lib.platforms.linux;
+    license = licenses.gpl2Plus;
+    maintainers = [ maintainers.raboof maintainers.carlthome ];
```
Comment (wegank) [inline pkgs/applications/audio/pd-plugins/gem/default.nix:48]:
```suggestion
    maintainers = with maintainers; [ raboof carlthome ];
```

File: pkgs/applications/audio/puredata/default.nix [hunk 2]
```diff
@@ -1,38 +1,61 @@
...
 
   meta = with lib; {
     description = ''A real-time graphical programming environment for
                     audio, video, and graphical processing'';
     homepage = "http://puredata.info";
     license = licenses.bsd3;
-    platforms = platforms.linux;
-    maintainers = [ maintainers.goibhniu ];
+    platforms = platforms.linux ++ platforms.darwin;
+    maintainers = [ maintainers.goibhniu maintainers.carlthome ];
```
Comment (wegank) [inline pkgs/applications/audio/puredata/default.nix:57]:
```suggestion
    maintainers = with maintainers; [ goibhniu carlthome ];
```

File: pkgs/applications/audio/pd-plugins/cyclone/default.nix [hunk 3]
```diff
@@ -2,13 +2,13 @@
...
-  version = "0.3beta-2";
+  version = "unstable-2023-09-12";
 
   src = fetchFromGitHub {
     owner = "porres";
     repo = "pd-cyclone";
-    rev = "cyclone${version}";
-    sha256 = "192jrq3bdsv626js1ymq10gwp9wwcszjs63ys6ap9ig8xdkbhr3q";
+    rev = "7c470fb03db66057a2198843b635ac3f1abde84d";
+    sha256 = "ixfnmeoRzV0qEOOIxCV1361t3d59fwxjHWhz9uXQ2ps=";
```
Comment (wegank) [inline pkgs/applications/audio/pd-plugins/cyclone/default.nix:11]:
```suggestion
    hash = "sha256-ixfnmeoRzV0qEOOIxCV1361t3d59fwxjHWhz9uXQ2ps=";
```

File: pkgs/applications/audio/pd-plugins/gem/default.nix [hunk 4]
```diff
@@ -10,17 +11,18 @@
...
+  version = "2023-07-28";
 
   src = fetchFromGitHub {
+
     owner = "umlaeute";
     repo = "Gem";
-    rev = "2edfde4f0587e72ef325e7f53681936dcc19655b";
-    sha256 = "0k5sq128wxi2qhaidspkw310pdgysxs47agv09pkjgvch2n4d5dq";
+    rev = "4ec12eef8716822c68f7c02a5a94668d2427037d";
+    sha256 = "Y/Z7oJdKGd7+aSk8eAN9qu4ss+BOvzaXWpWGjfJqGJ8=";
```
Comment (wegank) [inline pkgs/applications/audio/pd-plugins/gem/default.nix:25]:
```suggestion
   hash = "sha256-Y/Z7oJdKGd7+aSk8eAN9qu4ss+BOvzaXWpWGjfJqGJ8=";
```

File: pkgs/applications/audio/puredata/default.nix [hunk 5]
```diff
@@ -1,38 +1,61 @@
...
-stdenv.mkDerivation  rec {
+stdenv.mkDerivation rec {
   pname = "puredata";
-  version = "0.50-2";
+  version = "0.54-0";
 
   src = fetchurl {
     url = "http://msp.ucsd.edu/Software/pd-${version}.src.tar.gz";
-    sha256 = "0dz6r6jy0zfs1xy1xspnrxxks8kddi9c7pxz4vpg2ygwv83ghpg5";
+    sha256 = "6MFKfYV5CWxuOsm1V4LaYChIRIlx0Qcwah5SbtBFZIU=";
```
Comment (wegank) [inline pkgs/applications/audio/puredata/default.nix:20]:
```suggestion
    hash = "sha256-6MFKfYV5CWxuOsm1V4LaYChIRIlx0Qcwah5SbtBFZIU=";
```

**Review by magnetophon — APPROVED** [review]

**Review by wegank — APPROVED** [review]

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 230569` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>puredata</li>
  </ul>
</details>

**Comment by raboof** [comment]
> Thanks @raboof! I'll look into test running and fixing this when I get to a Linux workstation.
> 
> ```
> nix develop --impure --expr '{pkgs ? import <nixpkgs> {}, ...}: pkgs.mkShell { buildInputs = [ (pkgs.puredata-with-plugins [ pkgs.gem ]) ]; }'
> ```
> 
> Hoping it's just a matter of adding libndi to buildInputs.

Actually the NDI warning was just that, a warning. It seems something changed with the expected values passed to `-path`, but even with that fixed I'm now getting segmentation faults. Filed #261228 .

**Pull request merged**